presto-hive/src/main/java/io/prestosql/plugin/hive/MsckProcedure.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import io.prestosql.plugin.hive.metastore.Column;
+import io.prestosql.plugin.hive.metastore.Partition;
+import io.prestosql.plugin.hive.metastore.SemiTransactionalHiveMetastore;
+import io.prestosql.plugin.hive.metastore.Table;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.procedure.Procedure;
+import io.prestosql.spi.procedure.Procedure.Argument;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import static io.prestosql.plugin.hive.HiveMetadata.PRESTO_QUERY_ID_NAME;
+import static io.prestosql.plugin.hive.HivePartitionManager.extractPartitionValues;
+import static io.prestosql.plugin.hive.HiveSessionProperties.isRespectTableFormat;
+import static io.prestosql.plugin.hive.PartitionStatistics.empty;
+import static io.prestosql.plugin.hive.metastore.StorageFormat.fromHiveStorageFormat;
+import static io.prestosql.spi.block.MethodHandleUtil.methodHandle;
+import static io.prestosql.spi.type.StandardTypes.BOOLEAN;
+import static io.prestosql.spi.type.StandardTypes.VARCHAR;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class MsckProcedure
```

Comment:
They are ***almost*** equivalent.
* **Hive** DDL command takes two parameters: `repair` and  ` [ ADD | DROP | SYNC] `. It will update metastore if `repair` is set to true, otherwise just print out information about information about stale/missing partitions. `ADD` option will add any partitions that exist on HDFS but not in metastore to the metastore; `DROP` option will remove the partition that is already removed from file system from metastore; `SYNC` option will do both `ADD` and `DROP`.

* **Presto** `msck` procedure does the same thing in a more straightforward way. It takes two parameters for
  * `add`: add any partitions that exist on HDFS but not in metastore
  * `drop`: remove the partition that is already removed from file system from metastore

The combination of `add` and `drop` flag in Presto `msck` can achieve the same goal of `[ ADD | DROP | SYNC]` option in Hive.

The differences are:
- Presto does not print partition infomation if both `add` and `drop` are set to false. I should probably fix this so Presto msck also print out info.
- Hive support batching to avoid OOM when there's a large number of untracked partitions. I am not sure if this is necessary in Presto.

Reply:
Having a procedure return something to the client seems impossible. I looked at the `io.prestosql.execution.CallTask::execute` and it ignores the return value of `invokeWithArguments`, as shown at https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/execution/CallTask.java#L160.

@electrum